src/languages/javascript.js

```diff
@@ -63,29 +63,36 @@ export default function(hljs) {
     literal: ECMAScript.LITERALS.join(" "),
     built_in: ECMAScript.BUILT_INS.join(" ")
   };
-  const nonDecimalLiterals = (prefixLetters, validChars) =>
-    `\\b0[${prefixLetters}][${validChars}]([${validChars}_]*[${validChars}])?n?`;
-  const noLeadingZeroDecimalDigits = /[1-9]([0-9_]*\d)?/;
-  const decimalDigits = /\d([0-9_]*\d)?/;
-  const exponentPart = regex.concat(/[eE][+-]?/, decimalDigits);
+
+  // https://tc39.es/ecma262/#sec-literals-numeric-literals
+  const decimalDigits = '[0-9](_?[0-9])*';
+  const frac = `\\.(${decimalDigits})`;
+  // DecimalIntegerLiteral, including Annex B NonOctalDecimalIntegerLiteral
+  // https://tc39.es/ecma262/#sec-additional-syntax-numeric-literals
+  const decimalInteger = `0|[1-9](_?[0-9])*|0[0-7]*[89][0-9]*`;
   const NUMBER = {
     className: 'number',
     variants: [
-      { begin: nonDecimalLiterals('bB', '01') }, // Binary literals
-      { begin: nonDecimalLiterals('oO', '0-7') }, // Octal literals
-      { begin: nonDecimalLiterals('xX', '0-9a-fA-F') }, // Hexadecimal literals
-      { begin: regex.concat(/\b/, noLeadingZeroDecimalDigits, 'n') }, // Non-zero BigInt literals
-      { begin: regex.concat(/(\b0)?\./, decimalDigits, regex.optional(exponentPart)) }, // Decimal literals between 0 and 1
-      { begin: regex.concat(
-        /\b/,
-        noLeadingZeroDecimalDigits,
-        regex.optional(regex.concat(/\./, regex.optional(decimalDigits))), // fractional part
-        regex.optional(exponentPart)
-        ) }, // Decimal literals >= 1
-      { begin: /\b0[\.n]?/ }, // Zero literals (`0`, `0.`, `0n`)
+      // DecimalLiteral
+      { begin: `(\\b(${decimalInteger})((${frac})|\\.)?|(${frac}))` +
+        `[eE][+-]?(${decimalDigits})\\b` },
+      { begin: `\\b(${decimalInteger})\\b((${frac})\\b|\\.)?|(${frac})\\b` },
+
+      // DecimalBigIntegerLiteral
+      { begin: `\\b(0|[1-9](_?[0-9])*)n\\b` },
+
+      // NonDecimalIntegerLiteral
+      { begin: "\\b0[bB][0-1](_?[0-1])*n?\\b" },
+      { begin: "\\b0[oO][0-7](_?[0-7])*n?\\b" },
+      { begin: "\\b0[xX][0-9a-fA-F](_?[0-9a-fA-F])*n?\\b" },
```

Comment:
Who uses octal anyways?  Let's move hex and binary first in the sequencing... (perhaps a tiny perf improvement, but easy enough to do).

Reply:
This sequence matches the [spec](https://www.ecma-international.org/ecma-262/11.0/index.html#prod-NonDecimalIntegerLiteral) and I'm not sure what reordering you want; can you clarify and/or update on your own?